File: crates/mumble-tauri/ui/src/types/account.ts
/** Self-service account settings (own registration) and the events that
 *  deliver them. Backed by `FancyAccountSettings` (wire ID 154) /
 *  `FancyAccountAck` (156); operations are sent via the
 *  `update_account_settings` command (155 `FancyAccountSettingsUpdate`). */

/** Snapshot of the own registered account's server-side settings. */
export interface AccountSettings {
  /** True while the session belongs to a registered (non-SuperUser) account. */
  registered: boolean;
  /** Registered user ID of the account. */
  user_id?: number | null;
  /** Registered user name (DB casing). */
  name?: string | null;
  /** Contact email stored for the account. */
  email?: string | null;
  /** Whether password authentication is enabled. While set, a password is
   *  REQUIRED to log in - a certificate alone no longer works. */
  has_password: boolean;
  /** Whether a TOTP second factor is enrolled. */
  totp_enabled: boolean;
  /** Hex SHA-1 hash of the certificate bound to the account, if any. */
  cert_hash?: string | null;
  /** Whether this session's certificate matches `cert_hash` (precondition
   *  for switching back to certificate-only login). */
  cert_matches_session: boolean;
}

/** Actions understood by `update_account_settings` (snake_case names of
 *  `FancyAccountSettingsUpdate.Action`). */
export type AccountAction =
  | "query"
  | "set_password"
  | "clear_password"
  | "rename"
  | "set_email"
  | "unregister"
  | "totp_begin"
  | "totp_verify"
  | "totp_disable";

/** Numeric `FancyAccountSettingsUpdate.Action` values echoed in acks. */
export const ACCOUNT_ACTION_IDS: Record<AccountAction, number> = {
  query: 0,
  set_password: 1,
  clear_password: 2,
  rename: 3,
  set_email: 4,
  unregister: 5,
  totp_begin: 6,
  totp_verify: 7,
  totp_disable: 8,
};

/** Result of one account operation (`FancyAccountAck`). */
export interface AccountAck {
  /** Numeric `FancyAccountSettingsUpdate.Action` this ack belongs to. */
  action: number;
  ok: boolean;
  /** Machine-readable error code when `ok` is false (localised by the UI). */
  error?: string | null;
  /** TOTP enrolment: base32 shared secret for manual authenticator entry. */
  totp_secret?: string | null;
  /** TOTP enrolment: otpauth://totp/... provisioning URI. */
  totp_uri?: string | null;
}

/** Event payload emitted when a `FancyAccountSettings` snapshot arrives. */
export interface AccountSettingsEvent {
  settings: AccountSettings;
}

File: crates/mumble-tauri/src/state/types/account.rs
//! Self-service account settings types cached from `FancyAccountSettings` /
//! `FancyAccountAck` and surfaced to the "Account" settings panel.

use serde::Serialize;

/// Snapshot of the own registered account's server-side settings.
#[derive(Debug, Clone, Default, Serialize, serde::Deserialize, PartialEq, Eq)]
pub struct AccountSettings {
    /// True while the session belongs to a registered (non-SuperUser)
    /// account. False after a successful self-unregister.
    pub registered: bool,
    /// Registered user ID of the account.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub user_id: Option<u32>,
    /// Registered user name (DB casing).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    /// Contact email stored for the account.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub email: Option<String>,
    /// Whether password authentication is enabled. While set, a password is
    /// required to log in - a certificate alone no longer works.
    #[serde(default)]
    pub has_password: bool,
    /// Whether a TOTP second factor is enrolled.
    #[serde(default)]
    pub totp_enabled: bool,
    /// Hex SHA-1 hash of the certificate bound to the account, if any.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cert_hash: Option<String>,
    /// Whether the current session's certificate matches `cert_hash`
    /// (precondition for switching back to certificate-only login).
    #[serde(default)]
    pub cert_matches_session: bool,
}

/// Result of one account operation (`FancyAccountAck`).
#[derive(Debug, Clone, Serialize, serde::Deserialize, PartialEq, Eq)]
pub struct AccountAck {
    /// Echo of the `FancyAccountSettingsUpdate.Action` this ack belongs to.
    pub action: u32,
    pub ok: bool,
    /// Machine-readable error code when `ok` is false.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    /// TOTP enrolment: base32 shared secret for manual authenticator entry.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub totp_secret: Option<String>,
    /// TOTP enrolment: `otpauth://totp/...` provisioning URI.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub totp_uri: Option<String>,
}
